text/0032-improved-api-tokens.md

```diff
@@ -0,0 +1,153 @@
+- Start Date: 2022-10-27
+- RFC Type: feature
+- RFC PR: https://github.com/getsentry/rfcs/pull/32
+- RFC Status: draft
+
+---
+
+**Table of Contents**
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Why?](#why-)
+  - [Expected Outcome](#expected-outcome)
+    - [User Experience](#user-experience)
+    - [Token Format](#token-format)
+    - [Backend Token Hygiene](#backend-token-hygiene)
+    - [Secret Verification Service](#secret-verification-service)
+- [Background](#background)
+- [Options Considered](#options-considered)
+  - [Option #1](#option--1)
+  - [Option #2](#option--2)
+    - [Drawbacks](#drawbacks)
+  - [Option #3](#option--3)
+    - [Drawbacks](#drawbacks-1)
+- [Unresolved questions](#unresolved-questions)
+
+---
+
+# Summary
+
+Improve on Sentry's API token implementation to a more secure pattern. We will have three major goals:
+
+1. Using hashed values in the database
+2. Only display the token once to the end user upon creation
+3. Allow users to _name_ the tokens ([#9600](https://github.com/getsentry/sentry/issues/9600))
+4. Use a predictable prefix to integrate with various secret scanning services (ex. Github's Secret Scanning)
+
+# Motivation
+
+## Why?
+
+Sentry currently provides several strong options to secure a user's account, including SSO, SAML, and 2FA options. However our approach to handling API tokens is not as mature. We have two major issues with the current API token implementation:
+
+1. Tokens are stored as plaintext in the database, increasing the risk of exposure
+2. Tokens are visible as plaintext in the Sentry UI
+
+As a result, Sentry has to take extra steps to ensure the confidentially of API tokens (for example, [PR #39254](https://github.com/getsentry/sentry/pull/39254)).
+
+## Expected Outcome
+
+### User Experience
+
+When a user creates an API token, they **have the option to provide it an arbitrary name** and **the actual token value is only displayed to them once**.
+
+Existing API tokens, should be hashed and their original value discarded.
+
+A notice in the Sentry UI should be presented to suggest the user rotate and generate a new token to the new and improved version.
+
+### Token Format
+
+We need a predictable token format in order to integrate properly with secret scanning services. Our current format is a 64 character alphanumeric string. This is insufficient and would likely produce a high amount of false positives in tooling like [TruffleHog](https://github.com/trufflesecurity/trufflehog), [detect-secrets](https://github.com/Yelp/detect-secrets), [Github's Secret Scanning](https://docs.github.com/en/code-security/secret-scanning/about-secret-scanning), etc.
+
+The suggested pattern is `snty[a-zA-Z]_[a-zA-Z0-9]{64}`. The character _after_ `snty` will be used to identify the token type.
```

Comment:
At the risk of bikeshedding, my suggestion:  `\bsn[a-zA-Z]_[a-zA-Z0-9]{64}ty\b`

Reply:
I like the idea of the suffix, but I'm going to make some minor adjustments. 😄 

I'm still a fan of keeping the prefix as `snty`. This gives the opportunity for the regex to bail out early, and I'm making an assumption there's higher potential for more `sn`'s out there than `snty`. So we don't have two `ty`'s lol, I've opted to use `io` as our suffix.

If there are stronger feelings/reasonings about this, I'm glad to consider changing the format again.

Reply:
Opting for just the prefix.